File: priors/EventLocationPrior.py
# Copyright (c) 2012, Bayesian Logic, Inc.
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#     * Redistributions of source code must retain the above copyright
#       notice, this list of conditions and the following disclaimer.
#     * Redistributions in binary form must reproduce the above copyright
#       notice, this list of conditions and the following disclaimer in the
#       documentation and/or other materials provided with the distribution.
#     * Neither the name of Bayesian Logic, Inc. nor the
#       names of its contributors may be used to endorse or promote products
#       derived from this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
# Bayesian Logic, Inc. BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
# USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
# ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
# OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
# SUCH DAMAGE.
#
import os
import random
import numpy as np
from numpy import exp, log, pi, degrees, radians, arange, arcsin, zeros, sin
import time

from sigvisa.database.dataset import *
from sigvisa.utils.geog import dist_deg


from optparse import OptionParser



def kernel(bandwidth, distance):
    # convert bandwidth and distance to radians
    b, d = radians(bandwidth), radians(distance)

    return (1.0 + 1.0 / b ** 2)  / (2 * pi * AVG_EARTH_RADIUS_KM ** 2) \
        * exp(-d / b) / (1.0 + exp(-pi / b))


def compute_density(bandwidth, events, lon_arr, lat_arr, lon_grid, lat_grid):
    density = zeros((len(lon_arr), len(lat_arr)), float)
    for ev in events:
        density += kernel(bandwidth,
                          dist_deg((lon_grid, lat_grid),
                                   ev[[EV_LON_COL, EV_LAT_COL]]))
    density /= len(events)

    return density


def leave_oneout_avg_loglike(bandwidth, events, 
                             lon_arr, lat_arr,
                             lon_grid, lat_grid,
                             region_left,
                             bottom_range,
                             uniform_prob, earth_area,
                             lon_interval,
                             z_interval):
    print "computing leave one out bw %f" % bandwidth,
    t1 = time.time()
    n = float(len(events))
    density = compute_density(bandwidth, events, lon_arr, lat_arr, lon_grid, lat_grid)
    tot = 0.
    for ev in events:
        evdens = kernel(bandwidth, dist_deg((lon_grid, lat_grid),
                                            ev[[EV_LON_COL, EV_LAT_COL]]))
        netdensity = (n / (n - 1)) * density - (1 / (n - 1)) * evdens
        netdensity = (1 - uniform_prob) * netdensity + uniform_prob / earth_area

        loni = int((ev[EV_LON_COL] -region_left) // lon_interval)
        lati = int((sin(radians(ev[EV_LAT_COL])) -bottom_range) // z_interval)
        tot += log(netdensity[loni, lati])

    ans = tot / len(events)

    print "done (in %.1f secs) -> %f" % (time.time() - t1, ans)
    return ans


def learn(param_fname, options, leb_events, 
          sample_events=1000,
          uniform_prob=0.001,
          lon_interval=0.1,
          region_left=-180.,
          region_right=180.,
          region_top=90.,
          region_bottom=-90.):
    ########
    # first we will learn the optimal bandwidth (in degrees)
    ########

    # sample_events specifies how many events to use to compute the
    # optimal bandwidth

    # use half as many latitude buckets as longitude buckets


    lon_arr = arange(region_left, region_right, lon_interval)

    bottom_range = sin(radians(region_bottom))
    top_range = sin(radians(region_top))

    lon_coverage = (region_right - region_left)/360.
    lat_coverage = float(top_range - bottom_range) / 2.0

    n_lon_buckets = len(lon_arr)
    n_lat_buckets = int(n_lon_buckets * lat_coverage / (2* lon_coverage ))
    z_interval =  2.0 * lat_coverage / n_lat_buckets
    lat_arr = degrees(arcsin(arange(bottom_range, top_range, z_interval)))
    lat_grid, lon_grid = np.meshgrid(lat_arr, lon_arr)

    EARTH_AREA = 4 * pi * AVG_EARTH_RADIUS_KM ** 2
    solid_angle_sr = (top_range - bottom_range) * (radians(region_right) - radians(region_left))
    region_area = solid_angle_sr * EARTH_AREA / (4.*pi)
    patch_area = region_area / (len(lon_arr) * len(lat_arr))

    # to do so we will pick a random sample of 500 events
    indices = [i for i in range(len(leb_events))
               if random.random() < (float(sample_events) / len(leb_events))]

    # then we'll use leave-one-out cross-validation to select the best
    # bandwidth
    all_bandw, all_loglike = np.array([]), np.array([])
    for bandw in np.arange(0, 2, .2) + .2:
        all_bandw = np.append(all_bandw, bandw)
        all_loglike = np.append(all_loglike,
                                leave_oneout_avg_loglike(bandw,
                                                         leb_events[indices],
                                                         lon_arr=lon_arr,
                                                         lat_arr=lat_arr,
                                                         lon_grid=lon_grid, 
                                                         lat_grid=lat_grid,
                                                         region_left=region_left,
                                                         bottom_range=bottom_range,
                                                         uniform_prob=uniform_prob,
                                                         earth_area=region_area,
                                                         lon_interval=lon_interval,
                                                         z_interval=z_interval))

    best_bandw = all_bandw[all_loglike.argmax()]

    for bandw in np.arange(best_bandw - .2, best_bandw + .2, .05):
        if bandw == 0:
            continue
        if bandw in all_bandw:
            continue
        all_bandw = np.append(all_bandw, bandw)
        all_loglike = np.append(all_loglike,
                                leave_oneout_avg_loglike(bandw,
                                                         leb_events[indices],
                                                         lon_arr=lon_arr,
                                                         lat_arr=lat_arr,
                                                         lon_grid=lon_grid, 
                                                         lat_grid=lat_grid,
                                                         region_left=region_left,
                                                         bottom_range=bottom_range,
                                                         uniform_prob=uniform_prob,
                                                         earth_area=region_area,
                                                         lon_interval=lon_interval,
                                                         z_interval=z_interval))


    best_bandw = all_bandw[all_loglike.argmax()]

    print "best bandwidth", best_bandw

    """
    if options.gui:
        plt.figure(figsize=(8, 4.8))
        if not options.type1:
            plt.title("Leave-One-Out Avg. Log Likelihood")
        plt.scatter(all_bandw, all_loglike)
        plt.xlabel("bandwidth (degrees)")
        plt.ylabel("avg. log likelihood")
        plt.xlim(0, 2.)

        if options.writefig is not None:
            basename = os.path.join(options.writefig, "EventLocBandwidth")
            if options.type1:
                plt.savefig(basename + ".pdf")
            else:
                plt.savefig(basename + ".png")
    """

    density = compute_density(best_bandw, leb_events, 
                              lon_arr=lon_arr, lat_arr=lat_arr,
                              lon_grid=lon_grid, 
                              lat_grid=lat_grid)
    # fold-in a uniform prior
    density = (1 - uniform_prob) * density + uniform_prob / region_area


    if options.gui:
        title = "Event Location Log Density (b=%.2f)" % best_bandw

        f = Figure(figsize=(15, 15))
        hm = EventHeatmap(f=None, left_lon=region_left, right_lon=region_right, top_lat=region_top, bottom_lat=region_bottom, calc=False)
        ax = f.add_subplot(111)
        hm.init_bmap(axes=ax, nofillcontinents=True, projection="cyl")
        hm.plot_earth()
        
        lons = lon_arr
        lats = lat_arr

        #loni, lati = np.mgrid[0:len(lons), 0:len(lats)]
        #lon_arr, lat_arr = lons[loni], lats[lati]
        # convert to map coordinates
        x, y = hm.bmap(list(lon_arr.flat), list(lat_arr.flat))
        x_arr = np.array(x).reshape(lon_arr.shape)
        y_arr = np.array(y).reshape(lat_arr.shape)

        cs1 = hm.bmap.contour(x_arr, y_arr, vals, levels, linewidths=.5, colors="k",
                           zorder=6 - int(nolines))
        cs2 = hm.bmap.contourf(x_arr, y_arr, vals, levels, cmap=plt.cm.jet, zorder=5,
                               extend="both",
                               norm=matplotlib.colors.BoundaryNorm(cs1.levels,
                                                                   plt.cm.jet.N))


        plt.colorbar(cs2, orientation="vertical", drawedges=True)


        draw_density(bmap, lon_arr, lat_arr, log(density),
                     colorbar_orientation="horizontal", colorbar_shrink=0.75)
        if options.writefig is not None:
            basename = os.path.join(options.writefig, "EventLocationPrior")
            # a .pdf file would be too huge
            plt.savefig(basename + ".png")


    # convert the density at the grid points into a probability for each
    # bucket (the grid point is in the lower-left corner of the bucket)
    prob = density * patch_area

    print "Total prob:", prob.sum()

    fp = open(param_fname, "w")

    print >>fp, uniform_prob
    print >>fp, region_left, region_right, bottom_range, top_range
    print >>fp, lon_interval, z_interval

    np.savetxt(fp, prob)                 # writes out row by row
    fp.close()


class Dummy:
    pass


def main():


    parser = OptionParser()
    parser.add_option("--outfile", dest="outfile", default=os.path.join("parameters", "EventLocationPrior.txt"), type="str", help="")
    parser.add_option("--llnl_hack", dest="llnl_hack", default=False, action="store_true", help="")

    (options, args) = parser.parse_args()



    cursor = db.connect().cursor()
    param_fname = options.outfile
    if options.llnl_hack:
        print "reading events...",
        st1 = 1167609600
        et1 = 1199145600
        events1 = read_events(cursor, st1, et1, "isc")[0]

        st2 = 1203603300
        et2 = 1203624962
        events2 = read_events(cursor, st2, et2, "isc")[0]
        
        leb_events = np.vstack((events1, events2))


        region_left=-126.
        region_right=-100.
        region_bottom=32.
        region_top=49.
        uniform_prob=0.01
    else:
        start_time, end_time = read_timerange(cursor, "training", None, 0)
        print "reading events...",
        leb_events = read_events(cursor, start_time, end_time, "leb")[0]

        region_left=-180.
        region_right=180
        region_bottom=-90.
        region_top=90.
        uniform_prob=0.001

    print "done (%d events)" % len(leb_events)
    options = Dummy()

    options.gui = False
    options.type1 = False
    options.writefig = None

    learn(param_fname, options, leb_events, 
          uniform_prob=uniform_prob,
          region_left=region_left,
          region_right=region_right,
          region_bottom=region_bottom,
          region_top=region_top)


if __name__ == "__main__":
    main()

File: priors/EventLocationPrior.c
/*
 * Copyright (c) 2012, Bayesian Logic, Inc.
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *     * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *     * Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in the
 *       documentation and/or other materials provided with the distribution.
 *     * Neither the name of Bayesian Logic, Inc. nor the
 *       names of its contributors may be used to endorse or promote products
 *       derived from this software without specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
 * Bayesian Logic, Inc. BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
 * USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
 * OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 * 
 */

#include <stdlib.h>
#include <stdio.h>
#include <math.h>

#include "../sigvisa.h"

#define BUCKET(loni, lati) dist->p_bucketprob[(loni) * dist->numlat + (lati)]

void EventLocationPrior_Init_Params(EventLocationPrior_t * dist,
                                    const char * filename)
{
  FILE * fp;
  double uniform;
  int lonidx;
  int latidx;
  double cum_lon_prob;
  
  fp = fopen(filename, "r");

  if (!fp)
  {
    fprintf(stderr, "Unable to open file: %s\n", filename);
    exit(1);
  }

  if (1 != fscanf(fp, "%lg\n", &uniform))
  {
    fprintf(stderr, "error reading uniform prior from %s\n", filename);
    exit(1);
  }
  
  if (2 != fscanf(fp, "%lg %lg\n", &dist->lonstep, &dist->zstep))
  {
    fprintf(stderr, "error reading lonstep and zstep from %s\n", filename);
    exit(1);
  }

  dist->numlon = (int) (360.0 / dist->lonstep);
  dist->numlat = (int) (2.0 / dist->zstep);

  dist->p_bucketprob = (double *)calloc(dist->numlon * dist->numlat, 
                                        sizeof(*dist->p_bucketprob));
  
  for (lonidx = 0; lonidx < dist->numlon; lonidx ++)
  {
    for (latidx = 0; latidx < dist->numlat; latidx ++)
    {
      char * fmt;
      
      if (latidx == (dist->numlat-1))
        fmt = "%lg\n";
      else
        fmt = "%lg ";
      
      if (1 != fscanf(fp, "%lg ",
                      &dist->p_bucketprob[lonidx * dist->numlat + latidx]))
      {
        fprintf(stderr, "error reading %s at row %d col %d\n",
                filename, lonidx, latidx);
        exit(1);
      }
    }
  }

  fclose(fp);

  dist->p_lonprob = (double *)calloc(dist->numlon, sizeof(*dist->p_lonprob));
  dist->p_latprob = (double *)calloc(dist->numlon * dist->numlat,
                                     sizeof(*dist->p_latprob));

  cum_lon_prob = 0;
  for (lonidx = 0; lonidx < dist->numlon; lonidx ++)
  {
    dist->p_lonprob[lonidx] = 0;
    
    for (latidx = 0; latidx < dist->numlat; latidx ++)
    {
      dist->p_lonprob[lonidx] += dist->p_bucketprob[lonidx * dist->numlat 
                                                    + latidx];
      
    }

    for (latidx = 0; latidx < dist->numlat; latidx ++)
    {
      dist->p_latprob[lonidx * dist->numlat + latidx] 
        = dist->p_bucketprob[lonidx * dist->numlat + latidx]
        / dist->p_lonprob[lonidx];
    }

    cum_lon_prob += dist->p_lonprob[lonidx];
  }

  for (lonidx = 0; lonidx < dist->numlon; lonidx ++)
  {
    dist->p_lonprob[lonidx] /= cum_lon_prob;
  }

  /* initialize the probability of a hypothetical bucket at the north-pole,
   * this is useful for interpolating */
  dist->north_pole_prob = 0;
  for (lonidx = 0; lonidx < dist->numlon; lonidx ++)
    dist->north_pole_prob += dist->p_bucketprob[lonidx * dist->numlat 
                                                + dist->numlat-1];
  dist->north_pole_prob /= dist->numlon;
}

void EventLocationPrior_UnInit(EventLocationPrior_t * dist)
{
  free(dist->p_bucketprob);
  free(dist->p_lonprob);
  free(dist->p_latprob);
}

/* interpolate the value inside a rectangle given the values at the
 * four corners lower-left, lower-right, upper-left, upper-right and
 * the x, y values plus the x_size and y_size of the rectangle
 * (assuming 0,0 is lower-left)
 */
double interpolate_rect(double ll_val, double lr_val, double ul_val,
                        double ur_val, double x, double y,
                        double x_max, double y_max)
{
  double ll_wt, lr_wt, ul_wt, ur_wt;

  ll_wt = ((x_max - x) / x_max) * ((y_max - y) / y_max);
  lr_wt = (    x       / x_max) * ((y_max - y) / y_max);
  ul_wt = ((x_max - x) / x_max) * (    y       / y_max);
  ur_wt = (    x       / x_max) * (    y       / y_max);


  return (ll_wt * ll_val + lr_wt * lr_val + ul_wt * ul_val + ur_wt * ur_val)
    / (ll_wt + lr_wt + ul_wt + ur_wt);
}


double EventLocationPrior_LogProb(const EventLocationPrior_t * dist,
                                  double lon, double lat, double depth)
{
  double val;
  
  /* the event location better be valid */
  if ((lon < -180) || (lon > 180) || (lat < -90) || (lat > 90) 
      || (depth < MIN_DEPTH && depth != -1) || (depth > MAX_DEPTH))
  {
    printf("Error: invalid location -- lon %lf lat %lf depth %lf\n",
           lon, lat, depth);
    exit(1);
    
    return -HUGE_VAL;
  }

  if (lat == 90)
    val = dist->north_pole_prob;

  else
  {
    int lonidx, lonidx2;
    int latidx, latidx2;

    double lon_rem, z_rem;

    if (lon == 180)
      lon = -180;

    lonidx = (int) floor((lon - (-180)) / dist->lonstep);
    latidx = (int) floor((LAT2Z(lat) - (-1)) / dist->zstep);
    
    lon_rem = lon - (-180) - (lonidx * dist->lonstep);
    z_rem = LAT2Z(lat) - (-1) - (latidx * dist->zstep);

    lonidx2 = (lonidx + 1) % dist->numlon;
    
    if ((latidx+1) < dist->numlat)
    {
      latidx2 = latidx + 1;
    
      val = interpolate_rect(BUCKET(lonidx, latidx), BUCKET(lonidx2, latidx),
                             BUCKET(lonidx, latidx2), BUCKET(lonidx2, latidx2),
                             lon_rem, z_rem, dist->lonstep, dist->zstep);

    }
    else
    {
      val = interpolate_rect(BUCKET(lonidx, latidx), BUCKET(lonidx2, latidx),
                             dist->north_pole_prob, dist->north_pole_prob,
                             lon_rem, z_rem, dist->lonstep, dist->zstep);
    }
  }
  
  /* the probability density at any point in the bucket is uniform within the
   * bucket so we divide the bucket probability by the area of the bucket */
  double r = log(val) - log(dist->lonstep) - log(dist->zstep);
  if (depth == -1){
    return r;
  } else {
    return depth_lp(depth);
  }
}


double depth_lp(double depth) {

  // simple hack: depth prior is a mixture of
  //   .7 * Exponential(5)
  //   .3 * Gamma(loc=6.27209, alpha=1.41801048, scale=126.7144)
  //      (using the scipy.stats parameterization)
  // where the first component fits surface events and the second fits the tail of deep events.
  // See notebooks/depth_prior.ipynb for more.
  // (but also TODO: we should have more principled code for learning this, and especially 
  //  learn a joint location/depth prior since depth distribution is actually very different 
  //  in different locations). 

  if (depth < 0 || depth > 700) {
    printf("Error: invalid depth %lf\n", depth);    
    return -HUGE_VAL;
  }

  // first component
  double p1 = exp(-depth/5.0)/5.0;
    

  // second component
  double scale = 126.7144;
  double loc = 6.27209;
  double p2 = 0;
  
  if (depth > loc) {
    double x = (depth-loc)/scale;
    // gamma(1.41801048) = 0.88642998
    p2 = pow(x,0.41801048) * exp(-x) / 0.88642998;
    p2 /= scale;
  }

  // mixture
  double p = .7*p1 + .3*p2;
  return log(p);
}

static int sample_vec(int veclen, double * p_probvec)
{
  int i;
  double tgt;
  double cum;
  
  tgt = RAND_DOUBLE;
  
  cum = 0.0;
  for (i=0; (i<veclen); i++)
  {
    cum += p_probvec[i];

    if (cum >= tgt)
      break;
  }
  
  if (i >= veclen)
  {
    printf("sample_vec: tgt=%f cum=%f\n", tgt, cum);
    exit(1);
  }
  
  return i;
}

void EventLocationPrior_Sample(const EventLocationPrior_t * dist,
                               double * p_lon, double * p_lat, 
                               double * p_depth)
{
  int lonidx;
  int latidx;
  
  lonidx = sample_vec(dist->numlon, dist->p_lonprob);
  *p_lon = -180 + (lonidx + RAND_DOUBLE) * dist->lonstep;

  latidx = sample_vec(dist->numlat, dist->p_latprob + lonidx * dist->numlat);
  *p_lat = RAD2DEG * asin(-1 + (latidx + RAND_DOUBLE) * dist->zstep);

  *p_depth = RAND_UNIFORM(MIN_DEPTH, MAX_DEPTH);
  printf("WARNING: sampling depth from uniform distribution because I haven't implemented sampling for the actual depth prior yet.\n");
}

